src/stan_math_backend/Transform_Mir.ml

```diff
@@ -405,105 +757,187 @@ let add_reads vars mkread stmts =
      and then writes out the unconstrained versions *)
 let var_context_unconstrain_transform (decl_id, smeta, outvar) =
+  let st = outvar.Program.out_constrained_st in
   Stmt.Fixed.
     { pattern=
         Decl
-          { decl_adtype= UnsizedType.AutoDiffable
+          { decl_adtype=
+              UnsizedType.fill_adtype_for_type AutoDiffable
+                (SizedType.to_unsized st)
           ; decl_id
-          ; decl_type= Type.Sized outvar.Program.out_constrained_st
+          ; decl_type= Type.Sized st
           ; initialize= true }
     ; meta= smeta }
-  :: var_context_read (decl_id, smeta, outvar.Program.out_constrained_st)
-  @ [param_serializer_write ~unconstrain:true (decl_id, outvar)]
+  :: var_context_read (Stmt.Helpers.lvariable decl_id, smeta, st)
+  @ param_serializer_write ~unconstrain:true (decl_id, outvar)
 
 (** Reads in parameters from a serializer and then writes out the unconstrained versions *)
 let array_unconstrain_transform (decl_id, smeta, outvar) =
-  let out_constrained_st = outvar.Program.out_constrained_st in
-  [ Stmt.Fixed.
+  let decl =
+    Stmt.Fixed.
       { pattern=
           Decl
-            { decl_adtype= UnsizedType.AutoDiffable
+            { decl_adtype=
+                UnsizedType.fill_adtype_for_type AutoDiffable
+                  (SizedType.to_unsized outvar.Program.out_constrained_st)
             ; decl_id
-            ; decl_type= Type.Sized out_constrained_st
+            ; decl_type= Type.Sized outvar.Program.out_constrained_st
             ; initialize= true }
-      ; meta= smeta }
-  ; ( if SizedType.is_recursive_container out_constrained_st then
-      let nonarray_st, array_dims =
-        SizedType.get_array_dims out_constrained_st in
-      Stmt.Helpers.mk_nested_for (List.rev array_dims)
-        (fun loopvars ->
-          Stmt.Fixed.
+      ; meta= smeta } in
+  let rec read (lval, st) =
+    match st with
+    | SizedType.STuple subtypes ->
+        let elements =
+          List.mapi
+            ~f:(fun iter x ->
+              ((Stmt.Fixed.Pattern.LTupleProjection (lval, iter + 1), []), x) )
+            subtypes in
+        List.concat_map ~f:read elements
+    | _ when SizedType.contains_tuple st ->
+        let tupl, array_dims = SizedType.get_scalar_dims st in
+        [ Stmt.Helpers.mk_nested_for (List.rev array_dims)
+            (fun loopvars ->
+              Stmt.Fixed.
+                { meta= smeta
+                ; pattern=
+                    SList
+                      (read
+                         (let lbase, idxs = lval in
+                          ( ( lbase
+                            , idxs
+                              @ List.map
+                                  ~f:(fun e -> Index.Single e)
+                                  (List.rev loopvars) )
+                          , tupl ) ) ) } )
+            smeta ]
+    | _ when SizedType.is_recursive_container st ->
```

Comment:
This changes the meaning, since `SArray( SArray (SInt, 1), 1)` is a recursive container, but `SArray (SInt, 1)` isn't, so pulling out the subtype here would be incorrect.